exes/mqtt-broadcaster/inc/mqtt_broadcaster.hpp

```diff
@@ -0,0 +1,305 @@
+      return value;
+    } else if constexpr (std::is_arithmetic_v<value_type>) {
+      return std::to_string(value);
+    } else {
+      return "unknown";
+    }
+  }
+
+  // This function runs a coroutine for a slot, when a new value is received a message is sent to the MQTT broker
+  template <class slot_type, class value_type>
+  auto slot_coroutine(details::slot<slot_type>& slot, std::string signal_name) -> asio::awaitable<void> {
+    logger_.info("Starting coroutine for signal: {}", signal_name);
+    std::replace(signal_name.begin(), signal_name.end(), '.', '/');
+
+    while (!stop_coroutine_) {
+      std::expected<value_type, std::error_code> msg = co_await slot.async_receive(asio::use_awaitable);
+      if (msg) {
+        std::string message_value = convert_to_string(msg.value());
+        co_await send_message<value_type>(signal_name, message_value);
+      }
+    }
+  }
+
+  // This function sends a message to the MQTT broker. If the send fails then it will try to restart the broker
+  template <class value_type>
+  auto send_message(std::string signal, std::string value) -> asio::awaitable<void> {
+    logger_.trace("Sending message: {} to topic: {}", value, signal);
+    co_await asio::post(mqtt_client_->strand(), asio::use_awaitable);
+
+    bool send_failed = false;
+    try {
+      auto result = co_await mqtt_client_->send(
+          async_mqtt::v5::publish_packet{ mqtt_client_->acquire_unique_packet_id().value(),
+                                          async_mqtt::allocate_buffer(signal), async_mqtt::allocate_buffer(value),
+                                          async_mqtt::qos::at_least_once },
+          asio::use_awaitable);
+
+      if (result) {
+        logger_.error("Send failed: {}", result.message());
+        send_failed = true;
+      } else {
+        logger_.trace("Message sent successfully");
+      }
+    } catch (std::exception& e) {
+      logger_.error("Exception in send: {}", e.what());
+      send_failed = true;
+    }
+
+    if (send_failed) {
+      logger_.trace("Exception in send, reconnecting to broker");
+      if (!connect_active_) {
+        connect_active_ = true;
+        co_await asio::co_spawn(mqtt_client_->strand(), connect_to_broker(), asio::use_awaitable);
+      }
+    }
+    co_return;
+  }
+
+  std::string object_path_;
+  std::string interface_name_;
+  std::string mqtt_host_;
+  std::string mqtt_port_;
+  std::string mqtt_username_;
+  std::string mqtt_password_;
+
+  asio::io_context& ctx_;
+  ipc_client_type ipc_client_;
+  std::shared_ptr<mqtt_client_type> mqtt_client_;
+  tfc::logger::logger logger_;
+  std::unique_ptr<sdbusplus::asio::connection, std::function<void(sdbusplus::asio::connection*)>> dbus_connection_;
+  std::unique_ptr<sdbusplus::bus::match::match, std::function<void(sdbusplus::bus::match::match*)>> signal_updates_;
+  tfc::confman::config<config> config_;
+
+  std::vector<std::string> banned_signals_;
```

Comment:
it shouldn't be necessary to make a copy of the config, you can access the config vector from the config member

Reply:
I tried that initially, but the problem is that after a new config has arrived on the observable it isn't loaded immediatly. So when the observable is called:

```cpp
   // When a new list of allowed string arrives the program is restarted
    config_.value()._allowed_topics.observe([this]([[maybe_unused]] auto& new_conf, [[maybe_unused]] auto& old_conf) {
      restart();
    });
```

and consequently clean_signals is called:

```cpp
  // This function filter out signals that are not allowed
  auto clean_signals() -> void {
    active_signals_.erase(std::remove_if(active_signals_.begin(), active_signals_.end(),
                                         [&](const tfc::ipc_ruler::signal& signal) {
                                           return std::ranges::none_of(config_.value()._allowed_topics.value().begin(),
                                                                       config_.value()._allowed_topics.value().end(),
                                                                       [&](const std::string& allowed_string) {
                                                                         return signal.name.find(allowed_string) !=
                                                                                std::string::npos;
                                                                       });
                                         }),
                          active_signals_.end());
  }
```

the value of 

```cpp
config_.value()._allowed_topics.value().begin()
```

is still the old configuration. Then on the next update the previous one is broadcasted.

Reply:
I tried that initially, but the problem is that after a new config has arrived on the observable it isn't loaded immediatly. So when the observable is called:

```cpp
   // When a new list of allowed string arrives the program is restarted
    config_.value()._allowed_topics.observe([this]([[maybe_unused]] auto& new_conf, [[maybe_unused]] auto& old_conf) {
      restart();
    });
```

and consequently clean_signals is called:

```cpp
  // This function filter out signals that are not allowed
  auto clean_signals() -> void {
    active_signals_.erase(std::remove_if(active_signals_.begin(), active_signals_.end(),
                                         [&](const tfc::ipc_ruler::signal& signal) {
                                           return std::ranges::none_of(config_.value()._allowed_topics.value().begin(),
                                                                       config_.value()._allowed_topics.value().end(),
                                                                       [&](const std::string& allowed_string) {
                                                                         return signal.name.find(allowed_string) !=
                                                                                std::string::npos;
                                                                       });
                                         }),
                          active_signals_.end());
  }
```

the value of 

```cpp
config_.value()._allowed_topics.value().begin()
```

is still the old configuration. Then on the next update the previous one is broadcasted.

Reply:
you could use asio post

Reply:
after refactoring other parts of the code neither asio::post or this variable is necessary, simply accessing the value is enough.